spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,333 @@
+# Part 1: Background
+
+> **Warning**
+> We assume that you understand the Tendermint algorithm and therefore we will not review it here.
+If this is not the case, please refer to [here](../).
+
+Three kinds of messages are exchanged in the Tendermint algorithm: `PROPOSAL`, `PRE-VOTE`, and `PRE-COMMIT`.
+The algorithm progresses when certain conditions are satisfied over the set of messages received.
+For example, in order do decide on a value `v`, the set must include a `PROPOSAL` for `v` and more than two thirds of the number processes in `PRE-COMMIT` for the same `v`, during the same round.
+
+To ensure progress and termination, the algorithm assumes a **Global Stabilization Time (GST)**, after which communication is reliable and timely (Eventual $\Delta$-Timely Communication), which is used to provide **Gossip Communication**.
+
+| Eventual $\Delta$-Timely Communication|
+|-----|
+|There is a bound $\Delta$ and an instant GST (Global Stabilization Time) such that if a correct process $p$ sends a message $m$ at a time $t \geq \text{GST}$ to a correct process $q$, then $q$ will receive $m$ before $t + \Delta$.
```

Comment:
This means that a process can be correct from the perspective of one node in a network, and incorrect from the perspective of a different node in the same network. In other words, "correct" is not a property of the process itself, but is rather a property of a process relative to a given observer. Is that well understood?

Reply:
Being correct is a an absolute property of a node/process, not relative.
Nodes can wrongly suspect each other, but the definition of Gossip Communication does not take this into account.

Reply:
> Being correct is an absolute property of a node/process, not relative

From https://arxiv.org/pdf/1807.04938.pdf

> Formally, we model the network communication using a variant of the partially synchronous system model [6]: in all executions of the system there is a bound ∆ and an instant GST (Global Stabilization Time) such that all communication among correct processes after GST is reliable and ∆-timely, i.e., if a correct process p sends message m at time t ≥ GST to a correct process q, then q will receive m before t + ∆5. In addition to the standard partially synchronous system model [6], we assume an auxiliary property that captures gossip-based nature of communication6:

If "all communication among correct processes ... is reliable" within some upper bound (GST), and if processes are connected over a network, then "correctness" is necessarily subject to the network link(s) between processes. Those links are not equivalent for all processes, they're different for each connection: a given node may be reachable (correct) from one peer, but unreachable (incorrect) from another. As far as I can see, that's a  relative definition, until/unless those differences are resolved via e.g. consensus. 

With that said...

> Nodes can wrongly suspect each other, but the definition of Gossip Communication does not take this into account.

I guess this isn't the place to debate whether or not Gossip Communication is correct as defined :)

Reply:
A node may be correct and still be completely isolated from all the other nodes because its links to such nodes are faulty. I'd agree that, In practice, this node use useless, but it is still a correct node according to the model.

> I guess this isn't the place to debate whether or not Gossip Communication is correct as defined :)
💯